File: src/p1xxx/p1422.py
from leetcode_prelude import *


# Problem 1422
class Solution:
    def maxScore(self, s: str) -> int:
        total_zeros = s.count("0")

        left_zeros = 0
        right_ones = len(s) - total_zeros

        maximum = 0

        # dont visit the last item.
        # otherwise the right string will be empty in the last iteration.
        for ch in s[:-1]:
            if ch == "0":
                left_zeros += 1
            elif ch == "1":
                right_ones -= 1

            maximum = max(maximum, left_zeros + right_ones)

        return maximum

File: src/p1xxx/p1422.rs
// Problem 1422
impl Solution {
    pub fn max_score(s: String) -> i32 {
        let mut total_zeros = 0;
        let mut total_ones = 0;
        s.bytes().for_each(|s| match s {
            b'0' => total_zeros += 1,
            b'1' => total_ones += 1,
            _ => {}
        });

        let mut max = 0;
        let mut zeros = 0;
        let mut ones = 0;

        // substrings are non-empty
        // skip the last character
        s[..s.len() - 1].bytes().for_each(|s| {
            match s {
                b'0' => zeros += 1,
                b'1' => ones += 1,
                _ => {}
            }
            max = max.max(zeros + total_ones - ones);
        });

        max
    }
}
